### DISCUSSION TITLE
[Help] Note link autocompletion

### DISCUSSION THREAD

**Discussion by fmorroni** [zk-org/zk-nvim#176] [Help]
### Verify

- [X] I searched the existing discussions for help

### How can we help you?

Hi, I used to be able to get note links autocompletion by writing `[link text](())` where the note suggestions would appear inside the double parenthesis, and a pair of parenthesis would disappear when I chose a note to link to. This stopped working a while ago, probably (though not sure) when I updated to neovim 0.10.
Was this feature removed or is this a bug or a problem with my config (I didn't change anything from my config since updating though).
Autocompletion is still working by using double square brackets, but that sets the link text to the note's title, and the other way lets me set custom text (which I want most of the time). Of course I can just go back and change the text afterwards, but it's pretty bothersome.

**Comment by tjex** [comment]
Hello. I just checked on my end and I'm still getting completions.
You're sure nothing's changed with your `cmp` setup?

There was a PR addressing a bug with completion: https://github.com/zk-org/zk/pull/397

But as mentioned, it's still working on my end.

In your little code example you've written `[link](())`, i.e, placed 
a parentheses inside a parentheses. This won't return any files as the second 
parentheses is what `cmp` will use as completion.

It should look like this sequence `[link](` <- at this point, you should start 
to see completion popup.

**Reply by fmorroni** [reply]
Hi @tjex thanks for the answer, but I'm not sure what you mean. If I do `[link](` I only get buffer completions:
![image](https://github.com/zk-org/zk-nvim/assets/88512894/8373948f-e355-4767-aa41-cbb9a2762695)

This is how it looks when using `[[]]`:
![image](https://github.com/zk-org/zk-nvim/assets/88512894/26220999-b65a-489a-848c-c0b51be63974)
![image](https://github.com/zk-org/zk-nvim/assets/88512894/a075887c-6821-4c43-8b0b-367096fc0f0b)
![image](https://github.com/zk-org/zk-nvim/assets/88512894/f071fabd-3a75-48bc-b116-29538594e141)

And this previously worked as the above but letting me set any custom link text:
![image](https://github.com/zk-org/zk-nvim/assets/88512894/8c1af9f9-3402-4d70-887c-8213142134fa)

Are you saying that you get with `[link](` the same autocompletion as with `[[]]`?

**Reply by tjex** [reply]
ahhh. This will be because of your markdown link settings in the zk .config.

e.g, my settings:

```toml
# MARKDOWN SETTINGS
[format.markdown]

link-format = "markdown"
```

So when I write `[[]]` I also only get buffer completions, because zk lsp isn't seeing that you're wanting to hook into its functionality with a path completion.

So you would need to go in and change to `link-format = "markdown"` to get path completion settings when writing standard markdown links. 

(please mark as answer if the issue is solved 🙏)

**Reply by fmorroni** [reply]
Hi, I tried this but it had no effect. I know the config is working because if I change other things I do see the effects, but this setting didn't seem to change anything. I still get autocompletion from `[[]]` and nothing from `[text](` or `[text](())`

**Reply by tjex** [reply]
Ok. Would you mind posting here your entire .zk config, wrapped in code fences?

**Reply by tjex** [reply]
May as well post / link (to hosted dotfiles) your completion and zk config as well.

**Reply by fmorroni** [reply]
Hi @tjex sorry for the delay, here are my [dotfiles](https://github.com/fmorroni/dotfiles) and here's the link to the [zk config](https://github.com/fmorroni/dotfiles/tree/main/.config/zk).

**Reply by tjex** [reply]
@fmorroni really sorry. This was a lack of product experience on my side... Currently, only wikilinks are supported for autocompletion! See here [in the docs](https://github.com/zk-org/zk/blob/main/docs/editors-integration.md#language-server-protocol).

It was marksman that was offering me the completions when typing `[link]()`. 

So this leads me to believe something changed in your setup with your LSP, whereby `[link]()` completions stopped working completely. Did you uninstall a markdown LSP provider (such as marksman)?

Do you get completions when writing markdown docs in non-zk (where other files are within the tree to link to of course).  

In any case, the zk lsp never offered autocompletion for markdown links. So therein lies the confusion!

**Reply by mickael-menu** [reply]
That's not exactly right, `zk` does support auto-completion of Markdown links, but it is triggered as if you would write a Wiki Link (so with `[[`). I'm personally using Markdown links in my notebooks instead of WikiLinks. The [`link-format` setting](https://github.com/zk-org/zk/blob/main/docs/note-format.md#markdown) is used to know what type of link to generate.

However I don't think the double parenthesis completion you mentioned @fmorroni was ever part of `zk`. Maybe it was in another plugin you used?

**Reply by tjex** [reply]
ahhhhhh. I see now. But, then there's a bug. Or something odd is going on with my setup.
As when I select a completion option on any note that isn't from the initial completion list (i.e, after filtering), it still formats as wikilink, despite having `link-format = "markdown"`

This is why I thought it wasn't supported. Plus my expectations made me read [the doc](https://github.com/zk-org/zk/blob/main/docs/editors-integration.md#language-server-protocol) in a certain way.

**Reply by fmorroni** [reply]
> However I don't think the double parenthesis completion you mentioned @fmorroni was ever part of `zk`. Maybe it was in another plugin you used?

Hmm I guess if no one else remembers using double parenthesis it could have been another plugin on my side. But I have no clue what other plugin could have caused that, and I don't think I ever had another markdown language server besides the cli zk-ls. 
And could an external ls know about the zk notebook structure anyways? Because I was receiving the exact same auto-completions using double parenthesis than using double square brackets. It seems very odd to me that it would be a result of another plugin/ls.

In any case, thanks to both of you for your time.
If I find the time I may try to test the double parenthesis on an older version and a minimal config to check if they worked then.

**Reply by mickael-menu** [reply]
Actually you're right! I totally forgot I implemented this:

https://github.com/zk-org/zk/blob/725126575d00130c2fa423d36c89f0fff328a318/internal/adapter/lsp/server.go#L664-L665

And it does seem to be broken, I can't trigger it either with Nvim 0.10.0 and `zk` 0.14.1.

> ahhhhhh. I see now. But, then there's a bug. Or something odd is going on with my setup.
> As when I select a completion option on any note that isn't from the initial completion list (i.e, after filtering), it still formats as wikilink, despite having `link-format = "markdown"`

Are you using the LSP completion or a Neovim command like `ZkInsertLink`?

Unfortunately the LSP server is not covered by automated tests, so regressions are easy.

**Reply by tjex** [reply]
I'm using lsp completion via `nvim-cmp`. `ZkInsetLink` works as expected.

> And it does seem to be broken, I can't trigger it either with Nvim 0.10.0 and zk 0.14.1.

Alright, then something we need to look into for sure. As mentioned, we did address something similar, but it seems to be a larger issue at hand: [zk-org/zk#397](https://github.com/zk-org/zk/pull/397)

I'm also on that version combo.

**Reply by fmorroni** [reply]
> Actually you're right! I totally forgot I implemented this:
>
> https://github.com/zk-org/zk/blob/725126575d00130c2fa423d36c89f0fff328a318/internal/adapter/lsp/server.go#L664-L665

Okay after looking at this code it occurred to me to try the exact pattern shown in the code, and it actually works... If I use `]((` I get autocompletions. If I use `[]((` I also get autocompletions. It's only when I add the link text like this `[text]((` that it stops working.
![image](https://github.com/user-attachments/assets/96b62be8-410b-4e41-9f10-eaac0d8efcc7)
![image](https://github.com/user-attachments/assets/ea85a74b-f5a7-447a-bb2f-eb3334e37854)
![image](https://github.com/user-attachments/assets/d1c1eba3-0f0c-4546-b0f3-0eaf35411581)

Not sure if this is intended but I definitely could add any custom link text before updating.

It's also kinda buggy leaving the extra parenthesis after accepting an autocompletion:
![image](https://github.com/user-attachments/assets/f3437636-187a-4ca4-b533-f5e83388a0f3)

**Reply by mickael-menu** [reply]
I confirm that it's supposed to work with `[title]((`, this was the intended use case.

**Comment by tjex** [comment]
Reported: 

Markdown link completion:
https://github.com/zk-org/zk/issues/435

Incorrect link format insertion after cmp selection:
https://github.com/zk-org/zk/issues/436
